### PR TITLE
fix: Allow most basic and advanced gesture detectors together

### PR DESCRIPTION
**Pull request by spydon**
# Description

Most gesture detectors work together now, except for the `PanDetector` or `ScaleDetector` together with the `MultiTouchDragDetector`. Maybe even the `ScaleDetector` will work with the `MultiTouchDragDetector`, will have to try.

Also simplified the example game a little bit.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors and are passing (See [Contributor Guide]).
- [x] My PR does not decrease the code coverage, or I have __a very special case__ and explained on the PR description why this PR decreases the coverage.
- [x] I updated/added relevant documentation (doc comments with `///`) and updated/added examples in `doc/examples`.
- [x] I have formatted my code with `flutter format` and the `flutter analyze` does not report any problems.
- [x] I read and followed the [Flame Style Guide].
- [x] I have added a description of the change under `[next]` in `CHANGELOG.md`.
- [x] I removed the `Draft` status, by clicking on the `Ready for review` button in this PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flame users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in `CHANGELOG.md`).
- [x] No, this is *not* a breaking change.

## Related Issues

*Replace this paragraph with a list of issues related to this PR from the [issue database]. Indicate, which of these issues are resolved or fixed by this PR.*

[issue database]: https://github.com/flame-engine/flame/issues
[Contributor Guide]: https://github.com/flame-engine/flame/blob/main/CONTRIBUTING.md
[Flame Style Guide]: https://github.com/flame-engine/flame/blob/main/STYLEGUIDE.md

### CODE REVIEW

File: packages/flame/example/lib/main.dart [hunk 1]
```diff
@@ -14,14 +14,51 @@ void main() {
...
+  Future<void> onLoad() async {
+    await super.onLoad();
+    add(Square(Vector2(100, 200)));
+  }
+
+  @override
+  void onTapUp(int id, TapUpInfo info) {
+    super.onTapUp(id, info);
+    final touchPoint = info.eventPosition.game;
+    final handled = children.any((c) => c.shouldRemove);
```
Comment (luanpotter) [inline packages/flame/example/lib/main.dart:33]: interesting approach!

File: packages/flame/example/lib/main.dart [hunk 2]
```diff
@@ -14,14 +14,51 @@ void main() {
...
+
+    if (!handled) {
+      add(Square(touchPoint));
+    }
+  }
+
+  @override
+  void onDoubleTap() {
```
Comment (luanpotter) [inline packages/flame/example/lib/main.dart:41]: this seems like a very specific feature to showcase on this example

**Review by luanpotter — APPROVED** [review]

File: packages/flame/lib/src/components/mixins/tappable.dart [hunk 3]
```diff
@@ -61,6 +61,7 @@ mixin Tappable on Component {
   }
 }
 
+// TODO(spydon): these methods should return the result of propagateToChildren.
```
Comment (erickzanardo) [inline packages/flame/lib/src/components/mixins/tappable.dart:64]: Consider opening an issue for improved visibility

File: packages/flame/lib/src/game/game_widget/game_widget.dart [hunk 4]
```diff
@@ -264,34 +264,37 @@ class _GameWidgetState<T extends Game> extends State<GameWidget<T>> {
...
+    final hasBasicDetectors = hasBasicGestureDetectors(game);
+    final hasAdvancedDetectors = hasAdvancedGestureDetectors(game);
 
     assert(
-      !(hasBasicDetectors && hasAdvancedDetectors),
-      '''
-        WARNING: Both Advanced and Basic detectors detected.
-        Advanced detectors will override basic detectors and the later will not receive events
-      ''',
+      !(game is MultiTouchDragDetector &&
```
Comment (erickzanardo) [inline packages/flame/lib/src/game/game_widget/game_widget.dart:274]: Can we have some tests for these new assertions?

File: packages/flame/example/lib/main.dart [hunk 5]
```diff
@@ -14,14 +14,51 @@ void main() {
...
-class Square extends PositionComponent {
+/// This example simply adds a rotating white square on the screen, if you press
+/// somewhere other than on the existing square another square will be added and
+/// if you press on a square it will be removed.
+class MyGame extends FlameGame with DoubleTapDetector, HasTappables {
+  bool running = true;
+
+  @override
+  Future<void> onLoad() async {
+    await super.onLoad();
```
Comment (st-pasha) [inline packages/flame/example/lib/main.dart:25]: why do you need this, if super's `onLoad` is empty?

File: packages/flame/example/lib/main.dart (hunk 2 above)
Comment (spydon) [reply packages/flame/example/lib/main.dart:41]: It was here already, I just moved the class.

File: packages/flame/example/lib/main.dart (hunk 1 above)
Comment (spydon) [reply packages/flame/example/lib/main.dart:33]: I put a TODO on the onTap* that it should return whether any child handled the tap (returned false)

File: packages/flame/example/lib/main.dart (hunk 5 above)
Comment (spydon) [reply packages/flame/example/lib/main.dart:25]: It was already there, I just moved the class.
But I can remove it.

Comment (spydon) [reply packages/flame/example/lib/main.dart:25]: Ah right, it has @mustCallSuper on it.

Comment (st-pasha) [reply packages/flame/example/lib/main.dart:25]: well, the question then why must it, if the super doesn't do anything?

Comment (spydon) [reply packages/flame/example/lib/main.dart:25]: I don't remember, there was some reasoning behind it though. Do you remember @erickzanardo or @luanpotter?

File: packages/flame/lib/src/components/mixins/tappable.dart (hunk 3 above)
Comment (spydon) [reply packages/flame/lib/src/components/mixins/tappable.dart:64]: Opened #1211

File: packages/flame/example/lib/main.dart (hunk 5 above)
Comment (luanpotter) [reply packages/flame/example/lib/main.dart:25]: The reason is some FlameGame and Component mixins do initialization onLoad. so if people don't call super, those are missed, causing bugs. Thus we added the annoation so people would not make such mistake.

**Review by erickzanardo — APPROVED** [review]
LGTM!

**Comment by Liltreks** [comment]
Nice now we are outside any other suggests

**Pull request merged**